pkg/daemon/update.go

```diff
@@ -2026,3 +2029,163 @@ func (dn *Daemon) reboot(rationale string) error {
+	// We should never get here now though, since node-controller is waiting for the image
+	if desiredImage == "" {
+		glog.Infof("Looks like we don't have a desired image yet. Nothing to do.")
+		return nil
+	}
+
+	// I don't set currentImage ever, so we always hit the else block right now
+	if currentImage == desiredImage {
+		// Orrrr....if we've live updated to it
+		glog.Infof("Node is on proper image %s", desiredImage)
+
+		glog.Infof("Completing pending config %s", currentImage)
+		if err := dn.completeUpdate(currentImage); err != nil {
+			MCDUpdateState.WithLabelValues("", err.Error()).SetToCurrentTime()
+
+		}
+
+		// TODO(jkyros): For now I'm just making the pool happy so it's like "yeah I'm done"
+		if err := dn.nodeWriter.SetDone(dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name, desiredConfig); err != nil {
+			errLabelStr := fmt.Sprintf("error setting node's state to Done: %v", err)
+			MCDUpdateState.WithLabelValues("", errLabelStr).SetToCurrentTime()
+			return nil
+		}
+
+	} else {
+		// We think we have work to do
+
+		client := &RpmOstreeClient{}
+		state, err := client.GetState()
+		if err != nil {
+			return err
+		}
+
+		// Look through our deployments
+		// If we've rebased to it, but not booted, that's okay IF we've live-applied
+		for _, deployment := range state.Deployments {
+
+			// What we're looking for is at least in the list
+			if strings.TrimPrefix(deployment.ContainerImageReference, "ostree-unverified-registry:") == desiredImage {
+				// We rebased but we haven't booted, might be a liveapply
+				if deployment.Staged == true {
+
+					//TODO(jkyros): Check to see about liveapply rather than just marking this done
+					glog.Infof("Node is staged to %s, checking to see if we've liveapplied", desiredImage)
+					if err := dn.nodeWriter.SetDone(dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name, desiredConfig); err != nil {
+						errLabelStr := fmt.Sprintf("error setting node's state to Done: %v", err)
+						MCDUpdateState.WithLabelValues("", errLabelStr).SetToCurrentTime()
+						return nil
+					}
+
+					return nil
+				}
+
+				// Everything is perfect, we're already there the good way
+				if deployment.Booted == true {
+					glog.Infof("Node is already in image %s", desiredImage)
+					//TODO(jkyros): Someday actually set the current image
+					return nil
+				}
+			}
+
+		}
+
+		pullSecret, err := dn.getPullSecret()
+		if err != nil {
+			return err
+		}
+
+		//TODO(jkyros): what should these perms be
+		os.Mkdir("/run/ostree", 0544)
+
+		err = ioutil.WriteFile("/run/ostree/auth.json", pullSecret, 0400)
+		if err != nil {
+			return err
+		}
+
+		// I overwrite this in the image build now, so config checker still works AND I don't get stuck on it
+		//os.Remove(dn.currentConfigPath)
+
+		_, err = client.RebaseLayered(desiredImage)
+		if err != nil {
+			return err
+		}
+
+		// TODO(jkyros): This is a hack until we figure out file "owners"/layers/precedence
```

Comment:
See https://github.com/openshift/enhancements/pull/1032/commits/38b45871a19ea542d4d2f461db8f90027f344ea5 for some thoughts on this transition.